File: Stack/leetcode/20.go
package leetcode

func isValid(s string) bool {
	hash := map[byte]byte{')':'(', ']':'[', '}':'{'}

	stack := make([]byte, 0)
	if s == "" {
			return true
	}

	for i := 0; i < len(s); i++ {
			if s[i] == '(' || s[i] == '[' || s[i] == '{' {
					stack = append(stack, s[i])
			} else if len(stack) > 0 && stack[len(stack)-1] == hash[s[i]] {
					stack = stack[:len(stack)-1]
			} else {
					return false
			}
	}
	return len(stack) == 0
}

File: Stack/cpp/20.cpp
#include <stack>
#include <string>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        if (s.size() %2 != 0) {
            return false;
        }

        stack<char> st;

        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(') st.push(')');
            else if (s[i] == '{') st.push('}');
            else if (s[i] == '[') st.push(']');
            else if (st.empty() || st.top() != s[i]) return false;
            else st.pop();
        }
        return st.empty();
    }
};
